telefunc/node/server/shield/codegen/transformer.ts

```diff
@@ -0,0 +1,97 @@
+export { shieldTelefunctions }
+
+import { Project, VariableDeclarationKind } from 'ts-morph'
+
+function parseEmitMessages(messages: string[]) {
+  const matcher = /Type '"([A-Za-z0-9_]+)"' is not assignable to type '"([^"]+)"'./
+  const teleFuncToShieldStr: Record<string, string> = {}
+  
+  for (let message of messages) {
+    const found = message.match(matcher)
+    if (!found) {
+      console.warn("Encountered unrecognized emit message: ", message)
+      continue
+    }
+    if (found.length !== 3) {
+      console.warn(`Captured ${found.length - 1} groups with message: `, message)
+      continue
+    }
+
+    teleFuncToShieldStr[found[1]] = found[2]
+  }
+  return teleFuncToShieldStr
+}
+
+const shieldTelefunctions = (
+    telefuncSrc: string,
+    typesSrc: string  // contents of ./types.ts
+  ): string => {
+  const project = new Project({
+    compilerOptions: {
+      strict: true
+    }
+  })
+
+  project.createSourceFile("types.ts", typesSrc)
+  const telefuncSourceFile = project.createSourceFile("telefunc.ts", telefuncSrc)
+  // this source file is used for evaluating the template literal types' values
+  const shieldStrSourceFile = project.createSourceFile("shield-str.ts")
+
+  shieldStrSourceFile.addImportDeclaration({
+    moduleSpecifier: "./types",
+    namedImports: ["ShieldArrStr"]
+  })
+
+  const teleFunctions = telefuncSourceFile.getFunctions().filter(fun => fun.isExported())
+  const teleFunNames = teleFunctions.flatMap(fun => {
+    const name = fun.getName()
+    if (!name) return []
+    return [name]
+  })
+  shieldStrSourceFile.addImportDeclaration({
+    moduleSpecifier: "./telefunc",
+    namedImports: teleFunNames
+  })
+
+  // assign the template literal type to a string
+  // then diagnostics are used to get the value of the template literal type
+  for (const teleFunName of teleFunNames) {
+    shieldStrSourceFile.addVariableStatement({
+      declarations: [{
+        name: `${teleFunName}Shield`,
+        type: `ShieldArrStr<Parameters<typeof ${teleFunName}>>`,
+        initializer: `'${teleFunName}'`
+      }]
+    })
+  }
+
+  const diagnostics = shieldStrSourceFile.getPreEmitDiagnostics()
+  const messages = diagnostics.map(d => d.getMessageText().toString())
+  const teleFunToShieldString = parseEmitMessages(messages)
+
+  // TODO: do users ever want to add shield() calls themselves?
```

Comment:
Alternatively we can just add a collision avoiding prefix, so that the generated code is:

```js
import { shield as __shieldGenerator_shield } from 'telefunc'
const __shieldGenerator_t = __shieldGenerator_shield.type
````

Reply:
>   // TODO: do users ever want to add shield() calls themselves?

I'm also just realizing that not supporting this is a breaking change. But we can bump the semver major if needs be.

Reply:
What happens when calling `shield` multiple times for the same function?

Reply:
Not sure. I'll have to check. But we can show a warning, I guess that'd be enough to consider this not a breaking change.

Also one thing we can do is to flag the `shield()` call: `shield(fn, [t.string], { __generated: true })`. So that we can let the user's shield take over.

Reply:
I made it backward-compatible.